### PR TITLE
sctp: notify consumer if send buffer is full

### PR DESCRIPTION
**Pull request by jmillan**
And new maxSctpSendBufferSize transport option

### CODE REVIEW

File: src/Router.ts [hunk 1]
```diff
@@ -348,6 +348,7 @@ export class Router extends EnhancedEventEmitter
 			enableSctp = false,
 			numSctpStreams = { OS: 1024, MIS: 1024 },
 			maxSctpMessageSize = 262144,
+			maxSctpSendBufferSize = 262144,
```
Comment (ibc) [inline src/Router.ts:351]: This must also be added to plain and pipe transports.

File: src/WebRtcTransport.ts [hunk 2]
```diff
@@ -59,6 +59,12 @@ export type WebRtcTransportOptions =
 	 */
 	maxSctpMessageSize?: number;
 
+	/**
+	 * Maximum SCTP send buffer used by DataConsumers.
+	 * Default 262144.
+	 */
+	maxSctpSendBufferSize?: number;
```
Comment (ibc) [inline src/WebRtcTransport.ts:66]: same

**Review by ibc — CHANGES_REQUESTED** [review]

File: worker/src/RTC/SctpAssociation.cpp [hunk 3]
```diff
@@ -354,6 +367,11 @@ namespace RTC
 			  len,
 			  std::strerror(errno));
 		}
+
+		if (ret == EWOULDBLOCK || ret == EAGAIN)
+		{
+			Channel::Notifier::Emit(dataConsumer->id, "sctpsendbufferfull");
```
Comment (ibc) [inline worker/src/RTC/SctpAssociation.cpp:385]: This event is not handled in `DataConsumer.ts`.

**Review by ibc — CHANGES_REQUESTED** [review]

**Comment by jmillan** [comment]
NOTE:

````
	static size_t DefaultSctpSendBufferSize{ 262144 }; // 2^18.
	static size_t MaxSctpSendBufferSize{ 268435456 };  // 2^28.
````

I need to properly check those values. The first comes from the default value in my system, the second from a comment from [here]()https://github.com/sctplab/usrsctp/issues/142#issuecomment-294095656

File: worker/src/RTC/SctpAssociation.cpp (hunk 3 above)
Comment (ibc) [reply worker/src/RTC/SctpAssociation.cpp:385]: Done, pushed.

**Comment by ibc** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- lib/DataConsumer.js  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/versatica/mediasoup/pullRequest?prid=5829545&bid=19086333)

**Comment by ibc** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- lib/DataConsumer.js  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/versatica/mediasoup/pullRequest?prid=5829545&bid=19086333)

**Pull request merged**